File: src/Application Scheduler/Schedule Tests/SOROverloadCheck.cs
using System;
using System.Data;

namespace FSM
{
    public class SOROverloadCheck : ScheduleTest
    {
        protected override string TestName
        {
            get
            {
                return "SOR Overload";
            }
        }

        public override TestResult Test(DataRow testRow, int engineerID, DateTime day)
        {
            if (day != default)
            {
                int freeMins = 0;
                var maxTimes = App.DB.MaxEngineerTime.MaxEngineerTime_GetForEngineer(engineerID);
                if (maxTimes is object)
                {
                    var switchExpr = day.DayOfWeek;
                    switch (switchExpr)
                    {
                        case DayOfWeek.Monday:
                            {
                                freeMins = maxTimes.MondayValue;
                                break;
                            }

                        case DayOfWeek.Tuesday:
                            {
                                freeMins = maxTimes.TuesdayValue;
                                break;
                            }

                        case DayOfWeek.Wednesday:
                            {
                                freeMins = maxTimes.WednesdayValue;
                                break;
                            }

                        case DayOfWeek.Thursday:
                            {
                                freeMins = maxTimes.ThursdayValue;
                                break;
                            }

                        case DayOfWeek.Friday:
                            {
                                freeMins = maxTimes.FridayValue;
                                break;
                            }

                        case DayOfWeek.Saturday:
                            {
                                freeMins = maxTimes.SaturdayValue;
                                break;
                            }

                        case DayOfWeek.Sunday:
                            {
                                freeMins = maxTimes.SundayValue;
                                break;
                            }
                    }
                }

                var dtSOR = App.DB.Scheduler.SORTime_GetForEngineerAndDay(engineerID, day);
                int totalDaySOR = 0;
                if (dtSOR.Rows.Count > 0)
                {
                    totalDaySOR = Entity.Sys.Helper.MakeIntegerValid(dtSOR.Rows[0]["TotalDaySOR"]);
                }

                if (totalDaySOR + Entity.Sys.Helper.MakeIntegerValid(testRow["summedSOR"]) > freeMins)
                {
                    return new TestResult(string.Format("The SOR time total will overload this engineers time for the day"), false, true);
                }
                else
                {
                    return new TestResult();
                }
            }
            else
            {
                return new TestResult();
            }
        }
    }
}

File: src/Application Scheduler/Schedule Tests/SOROverloadCheck.vb
Public Class SOROverloadCheck
    Inherits ScheduleTest

    Protected Overrides ReadOnly Property TestName() As String
        Get
            Return "SOR Overload"
        End Get
    End Property

    Public Overrides Function Test(ByVal testRow As System.Data.DataRow,
                                   ByVal engineerID As Integer,
                                   ByVal day As Date) As ScheduleTest.TestResult

        If Not day = Nothing Then
            Dim freeMins As Integer = 0
            Dim maxTimes As Entity.MaxEngineerTimes.MaxEngineerTime = DB.MaxEngineerTime.MaxEngineerTime_GetForEngineer(engineerID)

            If Not maxTimes Is Nothing Then
                Select Case day.DayOfWeek
                    Case DayOfWeek.Monday
                        freeMins = maxTimes.MondayValue
                    Case DayOfWeek.Tuesday
                        freeMins = maxTimes.TuesdayValue
                    Case DayOfWeek.Wednesday
                        freeMins = maxTimes.WednesdayValue
                    Case DayOfWeek.Thursday
                        freeMins = maxTimes.ThursdayValue
                    Case DayOfWeek.Friday
                        freeMins = maxTimes.FridayValue
                    Case DayOfWeek.Saturday
                        freeMins = maxTimes.SaturdayValue
                    Case DayOfWeek.Sunday
                        freeMins = maxTimes.SundayValue
                End Select
            End If

            Dim dtSOR As DataTable = DB.Scheduler.SORTime_GetForEngineerAndDay(engineerID, day)
            Dim totalDaySOR As Integer = 0
            If dtSOR.Rows.Count > 0 Then
                totalDaySOR = Entity.Sys.Helper.MakeIntegerValid(dtSOR.Rows(0).Item("TotalDaySOR"))
            End If

            If totalDaySOR + Entity.Sys.Helper.MakeIntegerValid(testRow("summedSOR")) > freeMins Then
                Return New ScheduleTest.TestResult(String.Format("The SOR time total will overload this engineers time for the day"), False, True)
            Else
                Return New ScheduleTest.TestResult
            End If
        Else
            Return New ScheduleTest.TestResult
        End If

    End Function

End Class